.clang-tidy

```diff
@@ -43,6 +44,10 @@ readability-static-accessed-through-instance,
 readability-static-definition-in-anonymous-namespace,
 readability-string-compare,
 readability-uniqueptr-delete-release,
+-bugprone-exception-escape,
```

Comment:
We call py::print a lot in the destructors in tests which is quite noisy. Probably should nolint them all.

Reply:
How many nolints are needed?
Are they only under tests/?

Reply:
Quite a few, and yes only under tests. Under virtually ever destructor or operator that calls py::print.

Reply:
I'm thinking:
* We really want that check under include/
* But minimize clutter under tests/
Could we have a tests/.clang-tidy? How much trouble (copy-paste)?

Another completely different idea, possibly very naive/silly: could we make the prints noexcept?
```
void print() noexcept {
  try {
    whatever_we_have_already();
  }
  catch (...) {
    std::cout << "something useful" << std::endl;
    std::terminate();
  }
}
```

Reply:
The latter could definitely work and would fundamentally solve the problem. I think the compiler's main complaint is that you may not want to call std::terminate implicitly by throwing an exception in a destructor, but calling it explicitly should be fine. 

The former doesn't because we can only specify where clang-tidy is run by build taargets (ie. cpp) and since it's a header only library we can't ask it to run on headers unless they are actually used somewhere (and therefore have their templates expanded etc.)

Reply:
Alternatively, someone can just go an add all the NOLINT comments manually.

Reply:
> Alternatively, someone can just go an add all the NOLINT comments manually.

That sounds like a quick mostly mechanical effort, based on clang-tidy diagnostics, is that true? I'm thinking it would be best to do that straightaway in this PR, but it's fine with me to do it separately. Please let me know.

BTW: It looks like the PR description is out of date. Could you please update it to reflect the final state, if necessary?

Reply:
@rwgk Let's do it separetely. Having a no-except print version with std::terminate might be a better approach longterm.

Reply:
Sounds good.